### PR TITLE
Add API to remove items from stateful.Bundle

### LINKED ISSUE #136: Add API to remove items from `stateful.Bundle`
Hey, I'm testing an API that has add and remove functionality using `RuleBasedStateMachine`. 

E.g.

```python
class OrganisationAPI(RuleBasedStateMachine):
    orgs = Bundle('Organisation')
    users = Bundle('User')

    @rule(target=orgs, user=users, name=st.text(average_size=5))
    def create_org(self, user, name):
        return api.organisation_create(name, user)

    @rule(org=orgs, user=users)
    def delete_org(self, org, user):
        return api.organisation_delete(org, user)
```

Ideally the org bundle would not re-used deleted orgs. Filing as a reminder to myself, will investigate how to implement.

### PR DESCRIPTION
**Pull request by jomuel**
Implement the `consumes()` function proposed by @DRMacIver for this purpose.

Closes #136

### CODE REVIEW

File: hypothesis-python/src/hypothesis/stateful.py [hunk 1]
```diff
@@ -361,9 +384,15 @@ def rule(targets=(), target=None, **kwargs):
...
-    been produced for that bundle will be provided, if they are anything else
-    it will be turned into a strategy and values from that will be provided.
+    invocation. If their value is a Bundle, or if it is ``consumes(b)``
+    where ``b`` is a Bundle, then values that have previously been produced
+    for that bundle will be provided. If ``consumes`` is used, the value
+    will also be removed from the bundle.
+
+    Any other kwargs will be turned into a strategy and values from that
+    will be provided.
+
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/stateful.py]: nit: drop empty lines at end of docstring.

File: hypothesis-python/tests/cover/test_stateful.py [hunk 2]
```diff
@@ -352,6 +353,43 @@ def execute_step(self, step):
...
+    def depopulate_b1(self, consumed):
+        self.consumed_counter += 1
+        return consumed
+
+    @rule(value1=b1, value2=b2)
+    def check(self, value1, value2):
+        assert value1 != value2
+
+
+def test_consumes():
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_stateful.py]: Idiom: nstead of this function, we usually just assign `TestMachineWithConsumingRule = MachineWithConsumingRule.TestCase()` and let pytest discover the resulting class.

I'd also love to see a test or rule that uses collections of values, e.g. `lists(consumes(b1))`.

**Comment by jomuel** [comment]
Thanks, that's really nice. :smile: I'm happy to contribute to an innovative project like Hypothesis.

I addressed your comments except for the one with the additional test. When adding a rule like that I realized that using `consumes` within a strategy like `lists` does not work at the moment, as it would require `consumes` to return a `SearchStrategy`. I will look into how to make it work later, it should not be too hard.

**Comment by Zac-HD** [comment]
Perhaps the new I'll-call-it-a `BundleConsumer` class can inherit from `Bundle` to be a `SearchStrategy`, to simplify the drawing code?  I agree that this should be straightforward once you work out a nice way to do it.

And glad to hear you're enjoying contributing :grin:

**Comment by jomuel** [comment]
Good idea @Zac-HD, I made `Consumer` a subclass of `Bundle` and it seems to work.

File: hypothesis-python/src/hypothesis/stateful.py [hunk 3]
```diff
@@ -288,20 +290,39 @@ def do_draw(self, data):
         # Shrink towards the right rather than the left. This makes it easier
         # to delete data generated earlier, as when the error is towards the
         # end there can be a lot of hard to remove padding.
-        return bundle[cu.integer_range(data, 0, len(bundle) - 1, center=len(bundle))]
+        position = cu.integer_range(data, 0, len(bundle) - 1, center=len(bundle))
+        if self.consume:
+            return bundle.pop(position)
```
[the remark this answers is not in the document]
Comment (jomuel) [reply hypothesis-python/src/hypothesis/stateful.py]: I did not change this now as I didn't know when the sentinel values are supposed to be removed.

**Comment by jomuel** [comment]
Thanks for your reviews, I made the changes to the docstrings you both suggested. Now that I'm familiar with the stateful module already, I'd like to work on issue #321 too. :slightly_smiling_face:

File: hypothesis-python/src/hypothesis/stateful.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/stateful.py]: The idea is to never remove the sentinel values from the list but instead to leave them there and simply hide that fact when generating, but don't worry about it for now!

**Pull request merged**